src/ui/KonvergoWindow.cpp

```diff
@@ -458,18 +458,19 @@ void KonvergoWindow::updateWindowState(bool saveGeo)
     if (!isFullScreen() && saveGeo)
       saveGeometry();
 
-    setVisibility(QWindow::FullScreen);
-
+    setWindowFullScreen(true);
+#ifndef Q_OS_MAC
     // When fullscreening explicitly, we might have to move the window to a
```

Comment:
I'm not sure this is valid to ifdef that here. There are settings for the physicaly screen that the windows was last fullscreened on which is what `updateForcedScreen`will try to restore.
Native OSX fullscreen will indeed send it to a new desktop but that will remain on the same screen.
We should make sure that this behaves the same on OSX, so probably better to have the same scheme, if possible.
if you have 2 screens on OSX and decide PMP should be on second screen, the window has to restore fullscreen on second screen.

Reply:
I was thinking that it would be best to use the default macOS behavior, but I can make that change.

Reply:
yeah i think if we have options, they just need to be effective.

Reply:
Okay, I've been bashing my head against this, and I'm starting to get the feeling that it's not possible to implement this feature in macOS, at least not in a way that will work consistently. Fundamentally the issue is that full screen mode is a different beast in macOS than in any other OS. It has it's own standard behaviors and workflows, and one of the biggest fundamental differences is that apps are not supposed to have any control over which monitor they will occupy in full screen mode, nor are they supposed to have the ability to move between monitors. So we are very much going against the grain in implementing a feature like this.

So trying to control which monitor the window will use when it becomes full screen is unfortunately not a feature available in the macOS API. By and large a window will become full screen on whichever monitor it's on, but it's only guaranteed to work if the user drags the window to that monitor themselves.

If I use any of a variety of API calls to move the window to the correct monitor, *sometimes* macOS will fail to register that the window has changed screens. This has two bad effects: telling it to go full screen will of course send it to the wrong monitor, and worse yet, the monitor will be invisible with no easy way to bring it back. I cannot for the life of me figure out why this is happening, why it works sometimes but not other times, or whether there's a solution. It looks like a macOS window server bug, or more likely it might be something bad that Qt is doing in their implementation of `NSWindow`.

However, the one place this *does* seem to work consistently is when PMP starts up. So maybe it's worth keeping this in just so that on launch it will start on the correct display.

But there are a number cases where we simply cannot control which monitor it will use for full screen mode:

- The user can exit full screen mode using the title bar and drag the window to another monitor, as is normal in macOS
- The user can use Mission Control to move the window between monitors, as is normal in macOS
- If the user exits full screen mode via the app settings and then re-enables full screen mode later, we cannot ensure the window will appear on the correct monitor.
- Changing the "Force fullscreen display" setting generally does not successfully move the window to another monitor, though after rebooting PMP the window *will* appear on the correct monitor

I'm not sure how you think it's best to handle this. I'm thinking that in macOS PMP could respect the "force fullscreen display" setting at start up, but past that it will not try to force the window to appear on a particular display. Instead the user will have to move the window themselves via Mission Control or some other method (which is how things are normally done in macOS).

What do you think?

Reply:
Just to make this even more complicated, I just tried running the last build of PMP I made a few hours ago and suddenly "Force fullscreen display" is working correctly. Still not sure what causes it to get into a broken state, nor how often it happens!